File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.java
import java.util.List;
import java.util.ArrayList;
import java.util.Map;
import java.util.HashMap;

class Solution {
    List<List<String>> orders;
    List<String> answers;
    
    public List<String> letterCombinations(String digits) {
        String[] parsedDigits = digits.split("");
        Map<String, List<String>> phoneDictionary = new HashMap<>();
        phoneDictionary.put("2", List.of("a", "b", "c"));
        phoneDictionary.put("3", List.of("d", "e", "f"));
        phoneDictionary.put("4", List.of("g", "h", "i"));
        phoneDictionary.put("5", List.of("j", "k", "l"));
        phoneDictionary.put("6", List.of("m", "n", "o"));
        phoneDictionary.put("7", List.of("p", "q", "r", "s"));
        phoneDictionary.put("8", List.of("t", "u", "v"));
        phoneDictionary.put("9", List.of("w", "x", "y", "z"));
        
        orders = new ArrayList<>();
        answers = new ArrayList<>();
        
        if (digits.length() == 0) {
            return answers;
        }
        
        for(String s : parsedDigits) {
            orders.add(phoneDictionary.get(s));
        }
        
        int maxLength = orders.size();
        for (int i=0; i<maxLength; i++) {
            run(i, maxLength, "");
        }
        
        return answers;
    }
    
    public void run(int i, int maxLength, String tmp) {
        if (i == maxLength) {
            if (tmp.length() == maxLength) {
                answers.add(tmp);
            }
            return;
        }
        
        List<String> phoneDictionary = orders.get(i);
        for (int j=0; j<phoneDictionary.size(); j++) {
            run(i+1, maxLength, tmp + phoneDictionary.get(j));    
        }
    }
}

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.kt
class Solution {
    val orders = mutableListOf<List<String>>()
    val answers: MutableList<String> = mutableListOf<String>()
    
    fun letterCombinations(digits: String): List<String> {
        val phones = mutableMapOf<Char, List<String>>()
        val parsedDigits = digits.toCharArray()
        phones.put('2', listOf<String>("a","b","c"))
        phones.put('3', listOf<String>("d", "e", "f"))
        phones.put('4', listOf<String>("g", "h", "i"))
        phones.put('5', listOf<String>("j", "k", "l"))
        phones.put('6', listOf<String>("m", "n", "o"))
        phones.put('7', listOf<String>("p", "q", "r", "s"))
        phones.put('8', listOf<String>("t", "u", "v"))
        phones.put('9', listOf<String>("w", "x", "y", "z"))

        if (digits.length == 0) {
            return answers.toList()
        }
        
        for(s in parsedDigits) {
            orders.add(phones[s] ?: throw IllegalArgumentException("No Key"));
        }

        val maxLength = orders.size;
        for (i in 0..maxLength-1) {
            run(i, maxLength, "");
        }
        
        return answers.toList()
    }
    
    fun run(i: Int, maxLength: Int, tmp: String) {
        if (i == maxLength) {
            if (tmp.length == maxLength) {
                answers.add(tmp)
            }
            return
        }

        val phoneDictionary: List<String> = orders[i]
        for (j in 0..phoneDictionary.size-1) {
            run(i+1, maxLength, tmp + phoneDictionary[j])   
        }
    }
}